pkg/resource/deploy/step_generator.go

```diff
@@ -524,9 +525,14 @@ func (sg *stepGenerator) generateStepsFromDiff(
 			"unrecognized diff state for %s: %d", urn, diff.Changes)
 	}
 
-	// If there were changes, check for a replacement vs. an in-place update.
-	if diff.Changes == plugin.DiffSome {
-		if diff.Replace() {
+	// There are changes if the diff returned DiffSome or if there were initErrors previously that should
+	// force an update to occur even if no inputs changed.
+	needsUpdateOrReplace := diff.Changes == plugin.DiffSome || len(old.InitErrors) > 0
+
+	// If there were changes check for a replacement vs. an in-place update.
+	if needsUpdateOrReplace {
+		replaceKeys := sg.calculateReplaceKeys(old, new, diff)
```

Comment:
Factored this to compute replaceKeys in part as work toward being able to report back to users which keys are causing replacement in #6203.  For now, at least print that information explicitly in verbose diagnostics - as it is *very* useful for debugging customer issues.